pkg/daemon/update.go

```diff
@@ -975,6 +993,60 @@ func (dn *Daemon) updateOnClusterBuild(oldConfig, newConfig *mcfgv1.MachineConfi
 	return dn.reboot(fmt.Sprintf("Node will reboot into image %s / MachineConfig %s", newImage, newConfigName))
 }
 
+// Finalizes the revert process by enabling a special systemd unit prior to
+// rebooting the node.
+//
+// After we write the original factory image to the node, none of the files
+// specified in the MachineConfig will be written due to how rpm-ostree handles
+// file writes. Because those files are owned by the layered container image,
+// they are not present after reboot; even if we were to write them to the node
+// before rebooting. Consequently, after reverting back to the original image,
+// the node will lose contact with the control plane and the easiest way to
+// reestablish contact is to rebootstrap the node.
+//
+// By comparison, if we write a file that is _not_ owned by the layered
+// container image, this file will persist after reboot. So what we do is write
+// a special systemd unit that will rebootstrap the node upon reboot.
+// Unfortunately, this will incur a second reboot during the rollback process,
+// so there is room for improvement here.
+func (dn *Daemon) finalizeRevertToNonLayering(newConfig *mcfgv1.MachineConfig) error {
+	// First, we write the MachineConfig-encapsulated Ignition file. This is both
+	// the signal that the revert systemd unit should fire as well as the desired
+	// source of truth.
+	outBytes, err := json.Marshal(newConfig)
+	if err != nil {
+		return fmt.Errorf("could not marshal MachineConfig %q to JSON: %w", newConfig.Name, err)
+	}
+
+	if err := writeFileAtomicallyWithDefaults(constants.MachineConfigEncapsulatedPath, outBytes); err != nil {
```

Comment:
I'd like to propose we use a different file path for our revert unit, since:
1. there are other services that explicitly have references to this path, and I'm not sure if we actually want to do a full rebootstrap, and
2. the encapsulated json originally only contains content outside of ignition, so we're using this slightly differently

Just in case there are conflicts with any existing items

Reply:
I don't mind using a different path. I only used this particular path because I modeled the revert unit off of the original bootstrap unit.